plugins/ocp: Fix LM log command, Add set LM feature command, Add telemetry-log command

arthurshau opened the pull request:
Large PR which is meant to upstream the work I've done on Meta's internal OCP plugin (which is based off an older version of nvme-cli) to nvme-cli 2.3. 

Changes include:
- Reorganizing ocp-nvme.c to group related structs and functions
- Fixing typos in and updating the Latency Monitor Log functions to be OCP 2.0 compliant
- Adding the set latency monitor feature command
- Adding the telemetry log command

See commit messages for more details.

arthurshau commented:
@igaw Let me know if you want me to split the last two commits out into a separate PR since they're larger than the others.
@kdedow, @ikegami-t @lgdacunh Feel free to review if you'd like

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+
+		if (!((((tele_type == TELEMETRY_TYPE_HOST_0)
+				|| (tele_type == TELEMETRY_TYPE_HOST_1))
+				&& ((tele_area == 1) ||  (tele_area == 3)))
```
```diff
- 				&& ((tele_area == 1) ||  (tele_area == 3)))
+ 				&& ((tele_area == 1) || (tele_area == 3)))
```

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+
+		tele_area = cfg.area;
+
+		if (!((((tele_type == TELEMETRY_TYPE_HOST_0)
```
Indentation error and seems used uncessary `(` and `)` as complicated conditons and hard to understand.
```diff
- 		if (!((((tele_type == TELEMETRY_TYPE_HOST_0)
- 				|| (tele_type == TELEMETRY_TYPE_HOST_1))
- 				&& ((tele_area == 1) ||  (tele_area == 3)))
- 				|| (tele_type == TELEMETRY_TYPE_CONTROLLER && tele_area == 3))) {
+  		if (!(((tele_type == TELEMETRY_TYPE_HOST_0 || tele_type == TELEMETRY_TYPE_HOST_1) &&
+ 		    (tele_area == 1 || tele_area == 3)) || tele_type == TELEMETRY_TYPE_CONTROLLER &&
+ 		    tele_area == 3)) {
```

ikegami-t replied:
How aoubt the suggesution below?
```diff
- 		if (!(((tele_type == TELEMETRY_TYPE_HOST_0 || tele_type == TELEMETRY_TYPE_HOST_1) &&
- 		    (tele_area == 1 || tele_area == 3)) || tele_type == TELEMETRY_TYPE_CONTROLLER &&
- 		    tele_area == 3)) {
+  		if (tele_area != 1 || tele_area != 3 ||
+  		    tele_type == TELEMETRY_TYPE_CONTROLLER && tele_area != 3)
```

arthurshau replied:
I changed it to
```
if ((tele_area != 1 && tele_area != 3) ||
			(tele_type == TELEMETRY_TYPE_CONTROLLER && tele_area != 3))
```
instead.

ikegami-t replied:
I see. Thank you.

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -1,9 +1,9 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
-/* Copyright (c) 2022 Meta Platforms, Inc.
+/* Copyright (c) 2023 Meta Platforms, Inc.
```
```diff
- /* Copyright (c) 2023 Meta Platforms, Inc.
+ /*
+  * Copyright (c) 2023 Meta Platforms, Inc.
```

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -480,12 +541,142 @@ static int ocp_latency_monitor_log(int argc, char **argv,
+	};
+
+	err = nvme_set_features(&args);
+
```
Unnecessary line.

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+static void get_serial_number(struct nvme_id_ctrl *ctrl, char *sn)
+{
+	int i;
+	// Remove trailing spaces from the name
```
```diff
- 	// Remove trailing spaces from the name
+ 
+ 	/* Remove trailing spaces from the name */
```

ikegami-t reviewed (commented):
Just almost minor comments for the coding style.

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+	if (!cfg.type && !cfg.area) {
+		tele_type = TELEMETRY_TYPE_NONE;
+		tele_area = 0;
+	} else if ((!cfg.type + !cfg.area) == 0) {
```
Sorry I could understand as this works correctly but how about the following change as suggestion?
```diff
- 	} else if ((!cfg.type + !cfg.area) == 0) {
+ 	} else if (cfg.type && cfg.area) {
```

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -828,6 +828,434 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+	int err = 0;
+	const char *desc = "Retrieve and save telemetry log.";
+	const char *type = "Telemetry Type; 'host[Create bit]' or 'controller'";
+	const char *area = "Telemetry Data Area; 1 or 3";
```
Could you please let me know the reason to allow only the area 1 or 3 since it is described by the OCP documentation as below so seems the area 2 also should be used as option.
![tel-5](https://user-images.githubusercontent.com/44117581/230632235-78b6250a-2efa-45d8-b747-dd8c12a6cd4a.png)

igaw commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -828,6 +828,434 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Maybe trim the block length to 80 chars. Still pondering if we should split
the file if this is needed. IIRC I was the one who forced you to move all bits
into one file... So I can't really complain about this then, ay? ;)

igaw commented:
> @igaw Let me know if you want me to split the last two commits out into a separate PR since they're larger than the others. 

That's fine, it's new code, so I don't care too much about getting in smaller chunks. I am mostly interested in a good git history for bug fixes and refactoring parts that they are clean. It makes back porting etc way simpler. 

The only small grief I have are the `//` comment styles. I know I know, maybe we should just stop insisting on the awkward `/* */` style.

arthurshau commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+	struct nvme_dev *dev;
+	int err = 0;
+	const char *desc = "Retrieve and save telemetry log.";
+	const char *type = "Telemetry Type; 'host[Create bit]' or 'controller'";
```
I'm not sure - I'd have to ask the team. But I'm OK with keeping the type approach for now, just in case we expand it in the future.

arthurshau commented:
Hi all,

Sorry for the delay, these past few weeks have been pretty hectic for me. I believe I've addressed all of the comments here - for the larger functional changes involving the telemetry log command (such as adding Data Area 2 and changing the logic as suggested by @ikegami-t and @lgdacunh), I'll leave those for a later PR. I also plan to open a PR to add 2 more commands later to the plugin later this week.

@igaw Can you take a look at the failing builds? I don't think that's related to me, right?

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+		}
+
+		if (i != total_loop_cnt - 1) {
+			if (i == 0) {
```
Okay I see but sorry I missed to msetion as below as correctly. Anyway the behavior is not changed so I am okay with your change without any change.
```diff
-			if (i == 0) {
+			if (!i) {
```

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -638,6 +831,441 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+	else
+		last_xfer_size = transfersize;
+
+	if (filename == 0) {
```
{} are not necessary for the single statement.

ikegami-t commented on plugins/ocp/ocp-nvme.c:
```diff
@@ -638,6 +831,441 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+				__u32 data_len, void *data, __u8 nLSP, __u8 nRAE)
+{
+	struct nvme_passthru_cmd cmd = {
+		.opcode	= nvme_admin_get_log_page,
```
Seems indentation error as used tab character as thow spaces.

ikegami-t commented:
Reviewed-by: Tokunori Ikegami <ikegami.t@gmail.com>

arthurshau commented:
Addressed comments

igaw commented:
Thanks!
